device/src/keyboard.rs

```diff
@@ -0,0 +1,344 @@
+        is_active: bool,
+    ) {
+        let rect = key.rectangle;
+
+        if is_active {
+            rect.into_styled(
+                PrimitiveStyleBuilder::new()
+                    .stroke_color(Rgb565::WHITE)
+                    .stroke_width(1)
+                    .build(),
+            )
+            .draw(framebuf)
+            .unwrap();
+        } else {
+            rect.into_styled(
+                PrimitiveStyleBuilder::new()
+                    .stroke_color(Rgb565::new(5, 5, 5))
+                    .stroke_width(1)
+                    .build(),
+            )
+            .draw(framebuf)
+            .unwrap();
+
+            let font = U8g2TextStyle::new(fonts::u8g2_font_profont22_mf, Rgb565::WHITE);
+            Text::with_text_style(
+                key.label().to_string().as_str(),
+                rect.center(),
+                font,
+                TextStyleBuilder::new().alignment(Alignment::Center).build(),
+            )
+            .draw(framebuf)
+            .unwrap();
+        }
+    }
+
+    pub fn print_text_input(&mut self, framebuf: &mut FrameBuf<Rgb565, &mut [Rgb565; 67200]>) {
+        let mut x_offset = 0;
+        let mut y_offset = 0;
+        let spacing_size = 20;
+
+        let rect = Rectangle::new(
+            Point::new(x_offset, HEADER_BUFFER as i32),
+            Size::new(240, SCREEN_HEIGHT - HEADER_BUFFER - 2 * KEY_HEIGHT),
+        );
+
+        rect.into_styled(
+            PrimitiveStyleBuilder::new()
+                .fill_color(Rgb565::BLACK)
+                .build(),
+        )
+        .draw(framebuf)
+        .unwrap();
+
+        let chunked_backup =
+            self.buffer
+                .clone()
+                .into_iter()
+                .fold(vec![String::new()], |mut chunk_vec, char| {
+                    if chunk_vec.last().unwrap().len() < 4 {
+                        let last = chunk_vec.last_mut().unwrap();
+                        last.push(char);
+                    } else {
+                        chunk_vec.push(char.to_string());
+                    }
+                    chunk_vec
+                });
+
+        let textbox_style = TextBoxStyleBuilder::new().build();
+
+        // Don't show the top line once the backup gets to a certain length, "pan" down
+        if chunked_backup.len() <= 9 {
+            let _overflow = TextBox::with_textbox_style(
+                "frost1",
+                rect,
+                U8g2TextStyle::new(fonts::u8g2_font_profont22_mf, Rgb565::CYAN),
+                textbox_style,
+            )
+            .draw(framebuf)
+            .unwrap();
+            y_offset += spacing_size * 3 / 2;
+        }
+
+        // skip the first rows to only show the end 12 chunks
+        let rows_to_skip = if chunked_backup.len() <= 12 {
+            0
+        } else {
+            (chunked_backup.len() - 1) / 3 - 3
+        };
+
+        for (i, chunk) in chunked_backup[(rows_to_skip * 3)..].into_iter().enumerate() {
```

Comment:
Instead of drawing a textbox for each chunk, you can just add a space character after every 4 chars and pass it to one textbox

https://github.com/frostsnap/esp32-bsc/blob/cf6159c4dcebe067672d2b051e3cc29ef047bf53/frostsnap-graphics/src/lib.rs#L18